rtsp.go

```diff
@@ -675,26 +669,33 @@ func newRTSPCamera(ctx context.Context, _ resource.Dependencies, conf resource.C
 	}
 	cancelCtx, cancel := context.WithCancel(context.Background())
 	reader := gostream.VideoReaderFunc(func(_ context.Context) (image.Image, func(), error) {
-		latest := rc.latestOutput.Load()
-		if latest == nil {
-			return nil, func() {}, errors.New("no frame yet")
+		if videoCodec(rc.currentCodec.Load()) == MJPEG {
+			img := rc.latestMJPEGImage.Load()
+			if img == nil {
+				return nil, func() {}, errors.New("no mjpeg frame yet")
+			}
+			return *img, func() {}, nil
 		}
 
-		release := func() {}
-		// When model is mjpeg, latest frame wrapper will be nil.
-		if latest.frameWrapper != nil {
-			latest.frameWrapper.isBeingServed.Store(true)
+		frame := rc.latestFrame.Load()
```

Comment:
I don't think the relationship between loading the latest frame and bumping the ref count is correct. What can happen when `handleLatestFrame` is called concurrently with this `VideoReaderFunc`?

Reply:
Thinking out loud here:

If `handleLatestFrame` grabs the underlying mutex first and swaps out the old frame for the new frame atomically, deref-ing and returning the old frame to the pool. This would cause `VideoReaderFunc` to wait for a sec then `load` to access the new frame— there should be no issues.

If the other way around happens, `VideoReaderFunc` loads in first, then `handleLatestFrame` swaps in a new value + derefs and puts the old one back in the pool, then said deref be part of a race condition with the ref in `VideoReaderFunc`.

If my understanding is correct, a possible solution would be to use a mutex for both critical sections? Are there alternatives I am not aware of

Reply:
Added a mutex